aptos-move/framework/aptos-stdlib/sources/cryptography/crypto_algebra.spec.move

```diff
@@ -1,90 +1,162 @@
     }
 
     spec eq_internal<S>(handle_1: u64, handle_2: u64): bool {
         pragma opaque;
+        ensures result == spec_eq_internal<S>(handle_1, handle_2);
     }
 
     spec hash_to_internal<S, H>(dst: &vector<u8>, bytes: &vector<u8>): u64 {
         pragma opaque;
+        ensures result == spec_hash_to_internal<S,H>(dst, bytes);
     }
 
     spec inv_internal<F>(handle: u64): (bool, u64) {
         pragma opaque;
+        ensures result_1 == spec_inv_internal_1<F>(handle);
+        ensures result_2 == spec_inv_internal_2<F>(handle);
     }
 
     spec mul_internal<F>(handle_1: u64, handle_2: u64): u64 {
         pragma opaque;
+        ensures result == spec_mul_internal<F>(handle_1, handle_2);
     }
 
     spec multi_pairing_internal<G1,G2,Gt>(g1_handles: vector<u64>, g2_handles: vector<u64>): u64 {
         pragma opaque;
+        ensures result == spec_multi_pairing_internal<G1,G2,Gt>(g1_handles, g2_handles);
     }
 
     spec multi_scalar_mul_internal<G, S>(element_handles: vector<u64>, scalar_handles: vector<u64>): u64 {
         pragma opaque;
+        ensures result == spec_multi_scalar_mul_internal<G, S>(element_handles, scalar_handles);
     }
 
     spec neg_internal<F>(handle: u64): u64 {
         pragma opaque;
+        ensures result == spec_neg_internal<F>(handle);
     }
 
     spec one_internal<S>(): u64 {
         pragma opaque;
+        ensures result == spec_one_internal<S>();
     }
 
     spec order_internal<G>(): vector<u8> {
         pragma opaque;
+        ensures result == spec_order_internal<G>();
     }
 
     spec pairing_internal<G1,G2,Gt>(g1_handle: u64, g2_handle: u64): u64 {
         pragma opaque;
+        ensures result == spec_pairing_internal<G1, G2, Gt>(g1_handle, g2_handle);
     }
 
     spec scalar_mul_internal<G, S>(element_handle: u64, scalar_handle: u64): u64 {
         pragma opaque;
+        ensures result == spec_scalar_mul_internal<G, S>(element_handle, scalar_handle);
     }
 
     spec serialize_internal<S, F>(handle: u64): vector<u8> {
         pragma opaque;
+        ensures result == spec_serialize_internal<S, F>(handle);
     }
 
     spec sqr_internal<G>(handle: u64): u64 {
         pragma opaque;
+        ensures result == spec_sqr_internal<G>(handle);
     }
 
     spec sub_internal<G>(handle_1: u64, handle_2: u64): u64 {
         pragma opaque;
+        ensures result == spec_sub_internal<G>(handle_1, handle_2);
     }
 
     spec upcast_internal<S,L>(handle: u64): u64 {
         pragma opaque;
+        ensures result == spec_upcast_internal<S, L>(handle);
     }
 
     spec zero_internal<S>(): u64 {
         pragma opaque;
-    }
-
+        ensures result == spec_zero_internal<S>();
+    }
+
+    spec schema AbortsIfCryptographyAlgebraEnabled {
+        aborts_if !std::features::spec_cryptography_algebra_enabled();
+    }
+
+    // ----------------
+    // Helper functions
+    // ----------------
+
+    spec fun spec_add_internal<S>(handle_1: u64, handle_2: u64): u64;
```

Comment:
The argument you are trying to make is that we proof correctness relative to arbitrary algebra. This requires some justification which need to be documented. For example, one could wonder if constraining the crypto algebra via some group axioms would make a difference how and what we are specifying.

Reply:
Modeling crypto_algebra requires more work so the current version has been moved out of this PR.

Reply:
A question. Since this function uses handles, does this depends on the internal state, which is required to be modeled?

Reply:
It has been removed